File: scripts/typescript-qualification.py
#!/usr/bin/env python3
"""Create and verify the exact TypeScript package qualification receipt."""

from __future__ import annotations

import hashlib
import json
import os
import subprocess
from pathlib import Path
import re
import sys
import time


MAX_RECEIPT_BYTES = 1_048_576
PACKAGES = (
    ("objects", "objects"),
    ("stream", "stream"),
    ("inference", "inference"),
    ("machines", "machines"),
    ("fs", "filesystem"),
    ("sdk", "sdk"),
)


def digest(path: Path) -> tuple[str, int]:
    checksum = hashlib.sha256()
    size = 0
    with path.open("rb") as source:
        while block := source.read(1_048_576):
            checksum.update(block)
            size += len(block)
    return checksum.hexdigest(), size


def expected_assets(root: Path) -> list[dict[str, object]]:
    assets: list[dict[str, object]] = []
    for asset_slug, directory in PACKAGES:
        manifest = json.loads(
            (root / "typescript" / "packages" / directory / "package.json").read_text(
                encoding="utf-8"
            )
        )
        version = manifest["version"]
        assets.append(
            {
                "asset": f"acyclic-labs-{asset_slug}-{version}.tgz",
                "name": manifest["name"],
                "version": version,
            }
        )
    return assets


def create(output: Path, source_sha: str, root: Path) -> None:
    if not re.fullmatch(r"(?:[0-9a-f]{40}|[0-9a-f]{64})", source_sha):
        raise RuntimeError("source commit must be a full lowercase Git object ID")
    packages = []
    expected_names = set()
    for package in expected_assets(root):
        archive = output / str(package["asset"])
        checksum, size = digest(archive)
        package.update({"sha256": checksum, "size": size})
        packages.append(package)
        expected_names.add(archive.name)
    observed_names = {path.name for path in output.glob("*.tgz")}
    if observed_names != expected_names:
        raise RuntimeError("qualification output does not contain exactly the six core archives")
    receipt = {"revision": 1, "source_commit": source_sha, "packages": packages}
    target = output / "QUALIFICATION.json"
    if target.exists():
        raise RuntimeError("qualification receipt already exists")
    target.write_text(
        json.dumps(receipt, sort_keys=True, separators=(",", ":")) + "\n",
        encoding="utf-8",
    )


def read_receipt(path: Path) -> dict[str, object]:
    with path.open("rb") as source:
        payload = source.read(MAX_RECEIPT_BYTES + 1)
    if len(payload) > MAX_RECEIPT_BYTES:
        raise RuntimeError("qualification receipt exceeds its size bound")
    receipt = json.loads(payload)
    if not isinstance(receipt, dict) or set(receipt) != {
        "revision",
        "source_commit",
        "packages",
    }:
        raise RuntimeError("qualification receipt schema is invalid")
    packages = receipt["packages"]
    if type(receipt["revision"]) is not int or receipt["revision"] != 1 or not isinstance(packages, list) or len(packages) != 6:
        raise RuntimeError("qualification receipt schema is invalid")
    expected_keys = {"asset", "name", "version", "sha256", "size"}
    names = set()
    for package in packages:
        if not isinstance(package, dict) or set(package) != expected_keys:
            raise RuntimeError("qualification receipt package is invalid")
        asset = package["asset"]
        if (
            not isinstance(asset, str)
            or not re.fullmatch(r"acyclic-labs-[a-z]+-[0-9A-Za-z.+-]+\.tgz", asset)
            or not isinstance(package["name"], str)
            or not isinstance(package["version"], str)
            or not isinstance(package["sha256"], str)
            or not re.fullmatch(r"[0-9a-f]{64}", package["sha256"])
            or type(package["size"]) is not int
            or package["size"] <= 0
            or asset in names
        ):
            raise RuntimeError("qualification receipt package is invalid")
        names.add(asset)
    return receipt


def verify(
    receipt_path: Path, source_sha: str, asset_name: str, archive: Path, root: Path
) -> None:
    receipt = read_receipt(receipt_path)
    if receipt["source_commit"] != source_sha:
        raise RuntimeError("qualification receipt belongs to another source commit")
    identities = {
        (item["asset"], item["name"], item["version"])
        for item in receipt["packages"]
    }
    expected = {
        (item["asset"], item["name"], item["version"])
        for item in expected_assets(root)
    }
    if identities != expected:
        raise RuntimeError("qualification receipt does not identify the six core packages")
    matches = [item for item in receipt["packages"] if item["asset"] == asset_name]
    if len(matches) != 1:
        raise RuntimeError("archive is absent from the qualification receipt")
    checksum, size = digest(archive)
    if matches[0]["sha256"] != checksum or matches[0]["size"] != size:
        raise RuntimeError("archive bytes differ from the qualified artifact")


def consumer(root: Path, bun: str) -> dict[str, object]:
    """Run one small consumer through package exports, Node, and the WASM core."""
    started = time.monotonic()
    launcher = (
        ["cmd", "/d", "/c", bun]
        if os.name == "nt" and Path(bun).suffix.lower() in {".cmd", ".bat"}
        else [bun]
    )
    commands = [
        [*launcher, "x", "tsc", "-b", "--force", "typescript/packages/sdk/tsconfig.json",
         "--pretty", "false"],
        [*launcher, "x", "tsc", "-p", "typescript/packages/sdk/consumer-tsconfig.json",
         "--pretty", "false"],
        [*launcher, "test", "typescript/packages/sdk/test/public-consumer.test.ts"],
    ]
    for command in commands:
        try:
            result = subprocess.run(
                command, cwd=root, capture_output=True, text=True, check=False, timeout=90,
            )
        except (OSError, subprocess.TimeoutExpired) as failure:
            return {
                "schema": 1,
                "passed": False,
                "elapsed_ms": int((time.monotonic() - started) * 1000),
                "error": str(failure)[-2000:],
            }
        if result.returncode != 0:
            return {
                "schema": 1,
                "passed": False,
                "elapsed_ms": int((time.monotonic() - started) * 1000),
                "error": (result.stderr + result.stdout)[-2000:],
            }
    return {
        "schema": 1,
        "passed": True,
        "elapsed_ms": int((time.monotonic() - started) * 1000),
        "consumer": "typescript/packages/sdk/test/public-consumer.test.ts",
        "surfaces": ["stream", "objects", "filesystem", "wasm"],
    }


def main() -> None:
    if len(sys.argv) == 3 and sys.argv[1] == "consumer":
        result = consumer(Path(__file__).resolve().parent.parent, sys.argv[2])
        print(json.dumps(result, separators=(",", ":")))
        raise SystemExit(0 if result["passed"] else 1)
    if len(sys.argv) == 4 and sys.argv[1] == "create":
        create(Path(sys.argv[2]), sys.argv[3], Path(__file__).resolve().parent.parent)
    elif len(sys.argv) == 6 and sys.argv[1] == "verify":
        verify(
            Path(sys.argv[2]),
            sys.argv[3],
            sys.argv[4],
            Path(sys.argv[5]),
            Path(__file__).resolve().parent.parent,
        )
    else:
        raise SystemExit(
            "usage: typescript-qualification.py create OUTPUT SOURCE_SHA | "
            "verify RECEIPT SOURCE_SHA ASSET ARCHIVE | consumer BUN"
        )
if __name__ == "__main__":
    main()

File: scripts/typescript-qualification.mjs
#!/usr/bin/env node

import { createHash } from "node:crypto";
import { existsSync, readFileSync, readdirSync, writeFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { spawnSync } from "node:child_process";

const MAX_RECEIPT_BYTES = 1_048_576;
const PACKAGES = [
  ["objects", "objects"],
  ["stream", "stream"],
  ["inference", "inference"],
  ["machines", "machines"],
  ["fs", "filesystem"],
  ["sdk", "sdk"],
];
const root = resolve(dirname(fileURLToPath(import.meta.url)), "..");

function fail(message) {
  throw new Error(message);
}

function digest(path) {
  const payload = readFileSync(path);
  return [createHash("sha256").update(payload).digest("hex"), payload.length];
}

function expectedAssets() {
  return PACKAGES.map(([assetSlug, directory]) => {
    const manifest = JSON.parse(readFileSync(join(root, "typescript", "packages", directory, "package.json"), "utf8"));
    return { asset: `acyclic-labs-${assetSlug}-${manifest.version}.tgz`, name: manifest.name, version: manifest.version };
  });
}

function create(output, sourceSha) {
  if (!/^(?:[0-9a-f]{40}|[0-9a-f]{64})$/.test(sourceSha)) fail("source commit must be a full lowercase Git object ID");
  const packages = expectedAssets().map(packageEntry => {
    const [sha256, size] = digest(join(output, packageEntry.asset));
    return { ...packageEntry, sha256, size };
  });
  const expectedNames = new Set(packages.map(item => item.asset));
  const observedNames = new Set(readdirSync(output).filter(name => name.endsWith(".tgz")));
  if (expectedNames.size !== observedNames.size || [...expectedNames].some(name => !observedNames.has(name))) {
    fail("qualification output does not contain exactly the six core archives");
  }
  const target = join(output, "QUALIFICATION.json");
  if (existsSync(target)) fail("qualification receipt already exists");
  writeFileSync(target, `${JSON.stringify({ revision: 1, source_commit: sourceSha, packages })}\n`);
}

function readReceipt(path) {
  const payload = readFileSync(path);
  if (payload.length > MAX_RECEIPT_BYTES) fail("qualification receipt exceeds its size bound");
  const receipt = JSON.parse(payload.toString("utf8"));
  if (!receipt || typeof receipt !== "object" || Array.isArray(receipt) || Object.keys(receipt).sort().join() !== "packages,revision,source_commit") {
    fail("qualification receipt schema is invalid");
  }
  if (receipt.revision !== 1 || !Array.isArray(receipt.packages) || receipt.packages.length !== 6) fail("qualification receipt schema is invalid");
  const names = new Set();
  for (const item of receipt.packages) {
    if (
      !item || typeof item !== "object" || Array.isArray(item)
      || Object.keys(item).sort().join() !== "asset,name,sha256,size,version"
      || typeof item.asset !== "string" || !/^acyclic-labs-[a-z]+-[0-9A-Za-z.+-]+\.tgz$/.test(item.asset)
      || typeof item.name !== "string" || typeof item.version !== "string"
      || typeof item.sha256 !== "string" || !/^[0-9a-f]{64}$/.test(item.sha256)
      || !Number.isInteger(item.size) || item.size <= 0 || names.has(item.asset)
    ) fail("qualification receipt package is invalid");
    names.add(item.asset);
  }
  return receipt;
}

function verify(receiptPath, sourceSha, assetName, archive) {
  const receipt = readReceipt(receiptPath);
  if (receipt.source_commit !== sourceSha) fail("qualification receipt belongs to another source commit");
  const identity = item => `${item.asset}\0${item.name}\0${item.version}`;
  const observed = new Set(receipt.packages.map(identity));
  const expected = new Set(expectedAssets().map(identity));
  if (observed.size !== expected.size || [...expected].some(item => !observed.has(item))) fail("qualification receipt does not identify the six core packages");
  const matches = receipt.packages.filter(item => item.asset === assetName);
  if (matches.length !== 1) fail("archive is absent from the qualification receipt");
  const [sha256, size] = digest(archive);
  if (matches[0].sha256 !== sha256 || matches[0].size !== size) fail("archive bytes differ from the qualified artifact");
}

function consumer(bun) {
  const started = performance.now();
  const commands = [
    ["install", "--frozen-lockfile"],
    ["x", "tsc", "-b", "--force", "typescript/packages/sdk/tsconfig.json", "--pretty", "false"],
    ["x", "tsc", "-p", "typescript/packages/sdk/consumer-tsconfig.json", "--pretty", "false"],
    ["test", "typescript/packages/sdk/test/public-consumer.test.ts"],
  ];
  for (const args of commands) {
    const command = process.platform === "win32" && /\.(?:cmd|bat)$/i.test(bun) ? "cmd" : bun;
    const commandArgs = command === "cmd" ? ["/d", "/c", bun, ...args] : args;
    const result = spawnSync(command, commandArgs, { cwd: root, encoding: "utf8", timeout: 90_000 });
    if (result.error || result.status !== 0) return { schema: 1, passed: false, elapsed_ms: Math.trunc(performance.now() - started), error: String(result.error ?? `${result.stderr}${result.stdout}`).slice(-2000) };
  }
  return { schema: 1, passed: true, elapsed_ms: Math.trunc(performance.now() - started), consumer: "typescript/packages/sdk/test/public-consumer.test.ts", surfaces: ["stream", "objects", "filesystem", "wasm"] };
}

const [command, ...args] = process.argv.slice(2);
if (command === "create" && args.length === 2) create(resolve(args[0]), args[1]);
else if (command === "verify" && args.length === 4) verify(resolve(args[0]), args[1], args[2], resolve(args[3]));
else if (command === "consumer" && args.length === 1) {
  const result = consumer(args[0]);
  console.log(JSON.stringify(result));
  process.exitCode = result.passed ? 0 : 1;
} else fail("usage: typescript-qualification.mjs create OUTPUT SOURCE_SHA | verify RECEIPT SOURCE_SHA ASSET ARCHIVE | consumer BUN");
